File: Difficulty/Medium/[8] String to Integer (atoi)/solution.cpp
class Solution {
public:
    int myAtoi(string str) {
        if (str.size() == 0 || str.find_first_not_of(' ') == string::npos)
            return 0;
        if (str.size() == 1 && !isdigit(str[0]))
            return 0;
        int i = 0;
        char flag = ' ';
        i = str.find_first_not_of(' ');
        if (!isdigit(str[i]) && str[i] != '-' && str[i] != '+')
            return 0;
        else if (str[i] == '-'){
            flag = '-';
            i += 1;
        }
        else if (str[i] == '+'){
            flag = '+';
            i += 1;
        }
        long res = 0;
        for(; i < str.size(); ++i){
            if (isdigit(str[i])){
                res *= 10;
                res += int(str[i]) - 48;
            }
            else
                break;
            if (res > INT_MAX)
                if (flag == '-')
                    return INT_MIN;
                else
                    return INT_MAX;
        }
        if (res)
            if (flag == '-')
                return -int(res);
            else
                return int(res);
        else
            return 0;          
    }
    
};

File: Difficulty/Medium/[8] String to Integer (atoi)/solution.java
class Solution {
    public int myAtoi(String str) {
        boolean minus = false;
        int i = 0, len = str.length();
        while (i < len && str.charAt(i) == ' ') i++;
        if (i < len) {
            if (str.charAt(i) == '-') {
                minus = true;
                i++;
            } else if (str.charAt(i) == '+') {
                i++;
            }
        }
        if (i == len) return 0;
        long res = myAtoiHelper(str, i, minus);
        return (int)res;
    }

    public long myAtoiHelper(String str, int i, boolean minus) {
        if (str.charAt(i) < '0' || str.charAt(i) > '9') return 0;
        long res = 0;
        int len = str.length();
        while (i < len) {
            char c = str.charAt(i++);
            if (c >= '0' && c <= '9') {
                res = res * 10 + (minus ? ('0' - c) : (c - '0'));
            } else {
                return res;
            }
            if (res > Integer.MAX_VALUE) {
                return Integer.MAX_VALUE;
            }
            if (res < Integer.MIN_VALUE) {
                return Integer.MIN_VALUE;
            }
        }
        return res;
    }
}